command/debug/debug.go

```diff
@@ -0,0 +1,613 @@
+		}
+		outputs["agent"] = agent
+	}
+
+	// Capture cluster members information, including WAN
+	if c.configuredTarget("cluster") {
+		members, err := c.client.Agent().Members(true)
+		if err != nil {
+			errors = multierror.Append(errors, err)
+		}
+		outputs["members"] = members
+	}
+
+	// Write all outputs to disk as JSON
+	for output, v := range outputs {
+		marshaled, err := json.MarshalIndent(v, "", "\t")
+		if err != nil {
+			errors = multierror.Append(errors, err)
+		}
+
+		err = ioutil.WriteFile(fmt.Sprintf("%s/%s.json", c.output, output), marshaled, 0644)
+		if err != nil {
+			errors = multierror.Append(errors, err)
+		}
+	}
+
+	return errors
+}
+
+// captureDynamic blocks for the duration of the command
+// specified by the duration flag, capturing the dynamic
+// targets at the interval specified
+func (c *cmd) captureDynamic() error {
+	successChan := make(chan int64)
+	errCh := make(chan error)
+	durationChn := time.After(c.duration)
+	intervalCount := 0
+
+	c.UI.Output(fmt.Sprintf("Beginning capture interval %s (%d)", time.Now().Local().String(), intervalCount))
+
+	// We'll wait for all of the targets configured to be
+	// captured before continuing
+	var wg sync.WaitGroup
+
+	capture := func() {
+		timestamp := time.Now().Local().Unix()
+
+		// Make the directory that will store all captured data
+		// for this interval
+		timestampDir := fmt.Sprintf("%s/%d", c.output, timestamp)
+		err := os.MkdirAll(timestampDir, 0755)
+		if err != nil {
+			errCh <- err
+		}
+
+		// Capture metrics
+		if c.configuredTarget("metrics") {
+			wg.Add(1)
+
+			go func() {
+				metrics, err := c.client.Agent().Metrics()
+				if err != nil {
+					errCh <- err
+				}
+
+				marshaled, err := json.MarshalIndent(metrics, "", "\t")
+				if err != nil {
+					errCh <- err
+				}
+
+				err = ioutil.WriteFile(fmt.Sprintf("%s/%s.json", timestampDir, "metrics"), marshaled, 0755)
+				if err != nil {
+					errCh <- err
+				}
+
+				// Sleep as other dynamic targets wait collect for the whole interv	al
+				time.Sleep(c.interval)
+
+				wg.Done()
+			}()
+		}
+
+		// Capture pprof
+		if c.configuredTarget("pprof") {
+			wg.Add(1)
+
+			go func() {
+				pprofOutputs := make(map[string][]byte, 0)
+
+				heap, err := c.client.Debug().Heap()
+				if err != nil {
+					errCh <- err
+				}
+				pprofOutputs["heap"] = heap
+
+				// Capture a profile with a minimum of 1s
+				// TODO should be min across the board
+				s := c.interval.Seconds()
+				if s < 1 {
+					s = 1
+				}
+
+				// This will block for the interval
+				prof, err := c.client.Debug().Profile(int(s))
+				if err != nil {
+					errCh <- err
+				}
+				pprofOutputs["profile"] = prof
+
+				gr, err := c.client.Debug().Goroutine()
+				if err != nil {
+					errCh <- err
+				}
+				pprofOutputs["goroutine"] = gr
+
+				// Write profiles to disk
+				for output, v := range pprofOutputs {
+					err = ioutil.WriteFile(fmt.Sprintf("%s/%s.prof", timestampDir, output), v, 0755)
```

Comment:
nit: 644 seems better for a file that's not an executable. (755 for dirs since they need to be executable to cd into them).